docclean/cycle_gan.py

```diff
@@ -0,0 +1,147 @@
+            cycled_y = self.generator_g(fake_x, training=True)
+
+            # same_x and same_y are used for identity loss.
+            same_x = self.generator_f(real_x, training=True)
+            same_y = self.generator_g(real_y, training=True)
+
+            disc_real_x = self.discriminator_x(real_x, training=True)
+            disc_real_y = self.discriminator_y(real_y, training=True)
+
+            disc_fake_x = self.discriminator_x(fake_x, training=True)
+            disc_fake_y = self.discriminator_y(fake_y, training=True)
+
+            # calculate the loss
+            gen_g_loss = self.generator_loss(disc_fake_y)
+            gen_f_loss = self.generator_loss(disc_fake_x)
+
+            total_cycle_loss = self.calc_cycle_loss(real_x, cycled_x) + self.calc_cycle_loss(real_y, cycled_y)
+
+            # Total generator loss = adversarial loss + cycle loss
+            total_gen_g_loss = gen_g_loss + total_cycle_loss + self.identity_loss(real_y, same_y)
+            total_gen_f_loss = gen_f_loss + total_cycle_loss + self.identity_loss(real_x, same_x)
+
+            disc_x_loss = self.discriminator_loss(disc_real_x, disc_fake_x)
+            disc_y_loss = self.discriminator_loss(disc_real_y, disc_fake_y)
+
+        # Calculate the gradients for generator and discriminator
+        generator_g_gradients = tape.gradient(total_gen_g_loss,
+                                              generator_g.trainable_variables)
+        generator_f_gradients = tape.gradient(total_gen_f_loss,
+                                              generator_f.trainable_variables)
+
+        discriminator_x_gradients = tape.gradient(disc_x_loss,
+                                                  discriminator_x.trainable_variables)
+        discriminator_y_gradients = tape.gradient(disc_y_loss,
+                                                  discriminator_y.trainable_variables)
+
+        # Apply the gradients to the optimizer
+        self.generator_g_optimizer.apply_gradients(zip(generator_g_gradients,
+                                                       generator_g.trainable_variables))
+
+        self.generator_f_optimizer.apply_gradients(zip(generator_f_gradients,
+                                                       generator_f.trainable_variables))
+
+        self.discriminator_x_optimizer.apply_gradients(zip(discriminator_x_gradients,
+                                                           discriminator_x.trainable_variables))
+
+        self.discriminator_y_optimizer.apply_gradients(zip(discriminator_y_gradients,
+                                                           discriminator_y.trainable_variables))
+
+    def train(self, dirty_images: tf.data.Dataset, clean_images: tf.data.Dataset, epochs: int = 50):
+        for epoch in tqdm.tqdm(range(epochs), leave=False):
+            for image_x, image_y in tqdm.notebook.tqdm(tf.data.Dataset.zip((dirty_images, clean_images)), leave=False):
```

Comment:
do you need `tqdm.notebook`? I guess if you want to run it in a notebook

Reply:
Fixed in 92ee6ee.